### PR TITLE
storage: close sockets on error

### LINKED ISSUE #523: Connections not cleaned up if disconnected mid-download?
See #522 for the sample code.

In that example, if a client disconnects mid-way through a download, the connection apparently doesn't get cleaned up, and we're left with N-1 available connections in the pool.

Can we make sure that if a connection is broken, we clean up the socket and toss it ?

/cc @robertdimarco

### PR DESCRIPTION
**Pull request by stephenplusplus**
Continuing from #530
Fixes #523 
- [x]  Tests

### CODE REVIEW

**Comment by stephenplusplus** [comment]
I signed it!

**Comment by ryanseys** [comment]
Why don't merge?

**Comment by stephenplusplus** [comment]
Needs tests!

**Comment by stephenplusplus** [comment]
@ryanseys tests are ready if you want to take a glance. The feature implementation is the same.

@jgeewax / @robertdimarco this is the last outstanding PR before the next release is ready.

File: test/storage/file.js [hunk 1]
```diff
@@ -370,95 +376,270 @@ describe('File', function() {
...
+          self.emit('error', error);
+        });
+      }
+      nodeutil.inherits(FakeFailedRequest, FakeRequest);
+      extend(FakeFailedRequest, FakeRequest);
+
+      return FakeFailedRequest;
+    }
+
+    it('should send query.generation if File has one', function(done) {
```
Comment (stephenplusplus) [inline test/storage/file.js]: Nothing changed, just moved this test.

**Comment by stephenplusplus** [comment]
Going to merge since the implementation was reviewed previously.

**Pull request merged**